examples/adaptive_crag/application.py

```diff
@@ -0,0 +1,537 @@
+        lancedb_results=filtered_results
+    )
+
+
+@action(reads=["query"], writes=["exa_search_keywords"])
+def extract_keywords_for_exa_search(
+    state: State, extraction_attempts: int = ATTEMPTS
+) -> tuple[dict[str, str], State]:
+    """
+    Extract 2-5 keywords from the user `query` for a web search.
+    These keywords will be used to search the internet for the answer. Instead of the full query.
+    """
+    query = state["query"]
+    try:
+        query = str(
+            ask_gemini.create(
+                messages=[user_message(exa_search_query_template(query=query))],  # type: ignore
+                response_model=ExaSearchKeywords,
+                max_retries=extraction_attempts,
+            )
+        )
+    except Exception as e:
+        print(f"Error in extract_keywords_for_web_search: {e}")
+    return {"exa_search_keywords": query}, state.update(exa_search_keywords=query)
+
+
+@action(reads=["exa_search_keywords", "docs_limit"], writes=["exa_search_results"])
+def search_exa(state: State) -> tuple[dict[str, list[str]], State]:
+    """
+    More details about the Exa API:
+        - https://docs.exa.ai/reference/getting-started
+        - https://github.com/exa-labs/exa-py
+    """
+    try:
+        exa_search_results = exa.search_and_contents(
+            query=state["exa_search_keywords"],
+            num_results=state["docs_limit"],
+            highlights=True,
+        )
+        exa_search_results = [res.highlights[0] for res in exa_search_results.results]
+    except Exception as e:
+        print(f"Error in search_exa: {e}")
+        exa_search_results = []
+    return {"exa_search_results": exa_search_results}, state.update(
+        exa_search_results=exa_search_results
+    )
+
+
+@action(
+    reads=["query", "lancedb_results", "exa_search_results", "chat_history"],
+    writes=["chat_history", "lancedb_results", "exa_search_results"],
+)
+def ask_assistant(
+    state: State, attempts: int = ATTEMPTS
+) -> tuple[dict[str, str], State]:
+    """
+    Combine the `chat_history`, `query`, `lancedb_results`, and `exa_search_results` to ask the assistant for an answer.
+    `chat_history`, `lancedb_results`, and `exa_search_results` are used as context for the assistant and can be empty lists.
+    """
+    query = state["query"]
+    lancedb_results = state.get("lancedb_results", [])
+    exa_search_results = state.get("exa_search_results", [])
+    messages = state["chat_history"]
+    context = lancedb_results + exa_search_results
+    if context:
+        context = "\n".join(context)
+        query = (
+            query.strip()
+            + "\n<additional_context>\n"
+            + context
+            + "\n</additional_context>"
+        )
+    messages.append({"role": "user", "content": query.strip()})
+    try:
+        response = ask_gemini.create(
+            messages=messages, response_model=str, max_retries=attempts
+        )
+    except Exception as e:
```

Comment:
Which exceptions are we looking for? MIght be nice to narrow down.

Reply:
We get "InstructorRetryException" from Instructor after the attempts have been exhausted. Should we specify that here?

Reply:
Yep I think that's a good call